music-quiz/src/views/app.js

```diff
@@ -1,24 +1,57 @@
+import { useState, useEffect } from "react";
 import HeaderNavbar from "../presenters/headerNavbarPresenter";
 import { Route, Routes } from "react-router-dom";
 import "bootstrap/dist/css/bootstrap.min.css";
-import Login from "../presenters/loginPresenter.js"
-import Register from "../presenters/registerPresenter.js"
-import CreatePlaylist from "../presenters/createPlaylistPresenter"
+import Login from "../presenters/loginPresenter.js";
+import Register from "../presenters/registerPresenter.js";
+import CreatePlaylist from "../presenters/createPlaylistPresenter.js";
+import Loading from "./loadingView.js";
 
 function App(props) {
-  return (
+  const [currentUser, setCurrentUser] = useState(props.model.currentUser);
+  const [isLoaded, setIsLoaded] = useState(false);
+
+  function componentCreated() {
+    function onObserverNotification() {
+      setCurrentUser(props.model.currentUser);
+    }
+
+    function onComponentTakeDown() {
+      props.model.removeObserver(onObserverNotification);
+    }
+
+    props.model.addObserver(onObserverNotification);
+
+    return onComponentTakeDown;
+  }
+
+  useEffect(componentCreated, []);
+
+  useEffect(() => {
+    // Wait 500 milliseconds before displaying page
+    setInterval(() => {
+      setIsLoaded(true);
+    }, 500);
+  }, []);
```

Comment:
This is a little bit of a hack because now an user that is not logged in will need to wait 500ms every time he/she visits the site. However, the actual login attempt will usually go much faster (or in the worse case much slower if Firebase is under heavy load). If you want to render our app only after an user has tried to login you should probably store a new variable in the model `loginAttemptComplete = true` and notify observers when this value is set to true and it should be triggered from here:


```
function updateModelFromFirebase(model) {
  firebase.auth().onAuthStateChanged((user) => {
    if (user) {
        // We are logged in (we probably want to save username/email in the model as well).
        model.setCurrentUser(user.uid)
        
        // Get playlist updates.
        firebase.database().ref(`${REF}/users/${model.currentUser}/playlists/`).on("child_added", (firebaseData) => {
            model.addPlaylist(firebaseData.val())
        })
    } else {
        // We are not logged in.
        model.setCurrentUser(null)
    }
    
    // ADD THIS CODE:
    // The initial login attempt should be complete now!
    model.setLoginAttemptComplete(true);
  });
  return;
}
``

However, I have not tested this... so it might not work :)

Reply:
Will test that! 👍